server/controllers/admin/roles.js

```diff
@@ -0,0 +1,211 @@
+      res.json(rows);
+    })
+    .catch(next)
+    .done();
+}
+
+function update(req, res, next) {
+  const role = req.body;
+  delete role.uuid;
+  const sql = ` UPDATE role SET ? WHERE uuid = ?`;
+
+  db.exec(sql, [role, db.bid(req.params.uuid)])
+    .then((rows) => {
+      res.json(rows);
+    })
+    .catch(next)
+    .done();
+}
+
+
+function remove(req, res, next) {
+  const binaryUuid = db.bid(req.params.uuid);
+
+  const sql = `DELETE FROM role WHERE uuid = ?`;
+  db.exec(sql, binaryUuid)
+    .then((rows) => {
+      res.json(rows);
+    })
+    .catch(next)
+    .done();
+}
+
+// affect permission to a specific role
+
+function affectPages(req, res, next) {
+  const data = req.body;
+
+  const unitIds = data.unit_ids;
+  const roleUuid = db.bid(data.role_uuid);
+
+  const msg = {
+    state : 1,
+    message : 'success',
+  };
+  const deleteFromRole = `DELETE FROM role_unit WHERE role_uuid = ? `;
+  const affectPage = `
+    INSERT INTO  role_unit(uuid, unit_id, role_uuid)
+    VALUES( ?, ?, ?)`;
+
+  db.exec(deleteFromRole, roleUuid)
+    .then(() => {
+      const t = [];
+
+      if (!unitIds) {
+        return;
+      }
+      if (_.isArray(unitIds)) {
+        unitIds.forEach((id, idx) => {
+          t[idx] = db.exec(affectPage, [db.uuid(), id, roleUuid]);
+        });
+      } else {
+        t[0] = db.exec(affectPage, [db.uuid(), unitIds, roleUuid]);
+      }
+      return q.all(t);
+    })
+    .then(() => {
+      res.json(msg);
+    })
+    .catch(next)
+    .done();
+}
+
+
+// retireves affected and not affected role by a user id
+function listForUser(req, res, next) {
+  const userId = req.params.user_id;
+  const projectId = req.params.project_id;
+
+  const sql = `
+    SELECT DISTINCT q.uuid, q.label, q.project_id , q.affected
+    FROM (
+
+    SELECT DISTINCT BUID(r.uuid) as uuid, r.label, r.project_id ,1 as affected
+    FROM role r
+    JOIN user_role as ur ON ur.role_uuid = r.uuid
+    WHERE ur.user_id = ?
+
+    UNION 
+    
+    SELECT DISTINCT BUID(r.uuid) as uuid, r.label, r.project_id , 0 as affected
+    FROM role r
+    WHERE r.uuid NOT IN ( 
+      select ro.uuid 
+      FROM role ro
+      JOIN user_role as ur ON ur.role_uuid = ro.uuid
+      WHERE ur.user_id = ?
+    )
+    ) as q
+    WHERE q.project_id =?
+    ORDER BY q.label
+    `;
+
+  db.exec(sql, [userId, userId, projectId])
+    .then((roles) => {
+      res.json(roles);
+    })
+    .catch(next)
+    .done();
+}
+
+// affect roles to a user
+// roles ares permissions
+function affectToUser(req, res, next) {
+  const data = req.body;
+  const roles_uuids = data.role_uuids;
+  const userId = data.user_id;
+
+  const msg = {
+    state : 1,
+    message : 'success',
+  };
+
+
+  const deleleUserRoles = `DELETE FROM user_role WHERE user_id=? `;
+  const addRole = ` INSERT INTO user_role SET ?`;
+
+  db.exec(deleleUserRoles, userId)
+    .then(() => {
+      if (!roles_uuids) {
+        return msg;
+      } else if (_.isArray(roles_uuids)) {
```

Comment:
This could also use the array trick that I showed you up above.  Like this:
```js
const roleUuids = [].concat(roles_uuids);

// this will be an array of arrays.  Each array will contain the 
// [uuid, role_uuid, and user_id] in that order.
const parameters => rolesUUids.map(roleUuid => [db.uuid(), db.bid(roleUuid), userId]);

return db.exec(addRoles, parameters);
```